packages/sdk/src/PreVerificationGasCalculator.ts

```diff
@@ -84,29 +133,79 @@ export class PreVerificationGasCalculator {
    * Value of the 'preVerificationGas' is the cost overhead that cannot be calculated precisely or accessed on-chain.
    * It depends on blockchain parameters that are defined by the protocol for all transactions.
    * @param userOp - the UserOperation object that may be missing the 'signature' and 'paymasterData' fields.
+   @param gasOptions - apply EIP-7623 gas calculation.
    */
   estimatePreVerificationGas (
-    userOp: Partial<UserOperation>
+    userOp: Partial<UserOperation>,
+    gasOptions: GasOptions
   ): number {
     const filledUserOp = this._fillUserOpWithDummyData(userOp)
-    return this._calculate(filledUserOp)
+    return this._calculate(filledUserOp, gasOptions)
   }
 
-  _calculate (userOp: UserOperation): number {
+  _calculate (userOp: UserOperation, gasOptions: GasOptions): number {
     const packedUserOp = arrayify(encodeUserOp(userOp, false))
     const userOpWordsLength = (packedUserOp.length + 31) / 32
-    const callDataCost = packedUserOp
-      .map(
-        x => x === 0 ? this.config.zeroByteGasCost : this.config.nonZeroByteGasCost)
-      .reduce(
-        (sum, x) => sum + x
-      )
-    const userOpDataWordsOverhead = userOpWordsLength * this.config.perUserOpWordGasOverhead
-
-    const userOpSpecificOverhead = callDataCost + userOpDataWordsOverhead + this.config.perUserOpGasOverhead
+    const tokenCount = this._countTokens(packedUserOp)
+
+    let callDataOverhead = 0
+    let perUserOpOverhead = this.config.perUserOpGasOverhead
+    if (bytesToHex(arrayify(userOp.callData)).startsWith(EXECUTE_USEROP_METHOD_SIG)) {
+      perUserOpOverhead += this.config.execUserOpGasOverhead +
+          this.config.execUserOpPerWordGasOverhead * userOpWordsLength
+    } else {
+      callDataOverhead += Math.ceil(hexDataLength(userOp.callData) / 32) * this.config.perUserOpWordGasOverhead
+    }
+
+    const userOpSpecificOverhead = perUserOpOverhead + callDataOverhead
     const userOpShareOfBundleCost = this.config.fixedGasOverhead / this.config.expectedBundleSize
 
-    return Math.round(userOpSpecificOverhead + userOpShareOfBundleCost)
+    const userOpShareOfStipend = this.config.transactionGasStipend / this.config.expectedBundleSize
+
+    if (this.config.useEip7623) {
```

Comment:
Everything inside this `if` should be a private function. Then the code here will look like

```
if (this.config.useEip7623) {
   rerturn this._goUseEip7623()
}
return this._don'tUseEip7623()

```

Reply:
this is going to be very hairy, and doesn't add much clarity: I already extracted the _eip7623transactionGasCost function outside.
notice the gas calculation requires passing down all those parameters - unless this helper method is an internal function...